crates/ruma-events/src/space/child.rs

```diff
@@ -69,19 +82,218 @@ pub struct HierarchySpaceChildEvent {
+    }
+}
+
+impl<'a> Ord for SpaceChildOrdFields<'a> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self.order, other.order) {
+            // Events with order are ordered before events without order.
+            (Some(_), None) => Ordering::Less,
+            (None, Some(_)) => Ordering::Greater,
+            (Some(self_order), Some(other_order)) => self_order.cmp(other_order),
+            (None, None) => Ordering::Equal,
+        }
+        .then_with(|| self.origin_server_ts.cmp(&other.origin_server_ts))
+        .then_with(|| self.state_key.cmp(other.state_key))
+    }
+}
+
+impl<'a> PartialOrd for SpaceChildOrdFields<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T> SpaceChildOrd for &T
+where
+    T: SpaceChildOrd,
+{
+    fn space_child_ord_fields(&self) -> SpaceChildOrdFields<'_> {
+        (*self).space_child_ord_fields()
+    }
+}
+
+impl SpaceChildOrd for OriginalSpaceChildEvent {
+    fn space_child_ord_fields(&self) -> SpaceChildOrdFields<'_> {
+        SpaceChildOrdFields::new(
+            self.content.order.as_deref(),
+            self.origin_server_ts,
+            &self.state_key,
+        )
+    }
+}
+
+impl SpaceChildOrd for RedactedSpaceChildEvent {
+    fn space_child_ord_fields(&self) -> SpaceChildOrdFields<'_> {
+        SpaceChildOrdFields::new(None, self.origin_server_ts, &self.state_key)
+    }
+}
+
+impl SpaceChildOrd for SpaceChildEvent {
+    fn space_child_ord_fields(&self) -> SpaceChildOrdFields<'_> {
+        match self {
+            StateEvent::Original(original) => original.space_child_ord_fields(),
+            StateEvent::Redacted(redacted) => redacted.space_child_ord_fields(),
+        }
+    }
+}
+
+impl SpaceChildOrd for OriginalSyncSpaceChildEvent {
+    fn space_child_ord_fields(&self) -> SpaceChildOrdFields<'_> {
+        SpaceChildOrdFields::new(
+            self.content.order.as_deref(),
+            self.origin_server_ts,
+            &self.state_key,
+        )
+    }
+}
+
+impl SpaceChildOrd for RedactedSyncSpaceChildEvent {
+    fn space_child_ord_fields(&self) -> SpaceChildOrdFields<'_> {
+        SpaceChildOrdFields::new(None, self.origin_server_ts, &self.state_key)
+    }
+}
+
+impl SpaceChildOrd for SyncSpaceChildEvent {
+    fn space_child_ord_fields(&self) -> SpaceChildOrdFields<'_> {
+        match self {
+            SyncStateEvent::Original(original) => original.space_child_ord_fields(),
+            SyncStateEvent::Redacted(redacted) => redacted.space_child_ord_fields(),
+        }
+    }
+}
+
+impl SpaceChildOrd for HierarchySpaceChildEvent {
+    fn space_child_ord_fields(&self) -> SpaceChildOrdFields<'_> {
+        SpaceChildOrdFields::new(
+            self.content.order.as_deref(),
+            self.origin_server_ts,
+            &self.state_key,
+        )
+    }
+}
+
+/// Helper trait to sort `m.space.child` events using using the algorithm for [ordering children
```

Comment:
This isn't a trait.
```suggestion
/// Helper struct to sort `m.space.child` events using using the algorithm for [ordering children
```
or
```suggestion
/// Helper type to sort `m.space.child` events using using the algorithm for [ordering children
```